consensus/types/src/fork_versioned_response.rs

```diff
@@ -4,65 +4,48 @@ use serde::{Deserialize, Deserializer, Serialize};
 use serde_json::value::Value;
 use std::sync::Arc;
 
-// Deserialize is only implemented for types that implement ForkVersionDeserialize
-#[derive(Debug, PartialEq, Clone, Serialize)]
-pub struct ExecutionOptimisticFinalizedForkVersionedResponse<T> {
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub version: Option<ForkName>,
-    pub execution_optimistic: Option<bool>,
-    pub finalized: Option<bool>,
-    pub data: T,
-}
-
-impl<'de, F> serde::Deserialize<'de> for ExecutionOptimisticFinalizedForkVersionedResponse<F>
-where
-    F: ForkVersionDeserialize,
-{
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        #[derive(Deserialize)]
-        struct Helper {
-            version: Option<ForkName>,
-            execution_optimistic: Option<bool>,
-            finalized: Option<bool>,
-            data: serde_json::Value,
-        }
-
-        let helper = Helper::deserialize(deserializer)?;
-        let data = match helper.version {
-            Some(fork_name) => F::deserialize_by_fork::<'de, D>(helper.data, fork_name)?,
-            None => serde_json::from_value(helper.data).map_err(serde::de::Error::custom)?,
-        };
-
-        Ok(ExecutionOptimisticFinalizedForkVersionedResponse {
-            version: helper.version,
-            execution_optimistic: helper.execution_optimistic,
-            finalized: helper.finalized,
-            data,
-        })
-    }
-}
-
 pub trait ForkVersionDeserialize: Sized + DeserializeOwned {
     fn deserialize_by_fork<'de, D: Deserializer<'de>>(
         value: Value,
         fork_name: ForkName,
     ) -> Result<Self, D::Error>;
 }
 
-// Deserialize is only implemented for types that implement ForkVersionDeserialize
+/// Deserialize is only implemented for types that implement ForkVersionDeserialize.
+///
+/// The metadata of type M should be set to `EmptyMetadata` if you don't care about adding fields other than
+/// version. If you *do* care about adding other fields you can mix in any type that implements
+/// `Deserialize`.
 #[derive(Debug, PartialEq, Clone, Serialize)]
-pub struct ForkVersionedResponse<T> {
+pub struct ForkVersionedResponse<T, M = EmptyMetadata> {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub version: Option<ForkName>,
+    #[serde(flatten)]
+    pub metadata: M,
     pub data: T,
 }
 
-impl<'de, F> serde::Deserialize<'de> for ForkVersionedResponse<F>
+/// Metadata type similar to unit (i.e. `()`) but deserializes from a map (`serde_json::Value`).
+///
+/// Unfortunately the braces are semantically significant, i.e. `struct EmptyMetadata;` does not
```

Comment:
Curious why this is. I tried compiling it without the braces and it seemed to work, so didn't understand why.

Reply:
Yeah it's crazy, the tests all fail when you use the `;` form, because `serde` insists that only "maps and structs" may be used with flatten (and "unit structs" don't count). So _serialisation_ actually fails, and you get really weird 500 errors from Warp, which helpfully [_logs_ the serialisation error using `log` and then returns an empty response](https://docs.rs/warp/latest/warp/reply/fn.json.html).

Reply:
Wow crazy! Thanks for explaining :)